Add hook parameter

jim-barber-he opened the pull request:
Add the ability to run this as either an environment hook or pre-command hook with the former being the default.
The behaviour is chosen via a new hook parameter.

I tried switching from the cultureamp/aws-assume-role plugin to this one but found steps using the artifacts plugin were failing. It is because this plugin was running before the artifacts had a chance to run, and so the role being assumed was not able to access our artifacts S3 bucket. The plugin we switched from ran as a pre-command hook and didn't suffer from this problem.

Move the code for the plugin from hooks/environment to a shared lib/plugin.bash script.
Update hooks/environment to call the shared code if the new hook parameter is unset (making it the default) or set to environment. Create hooks/pre-command to call the shared code if the new hook parameter is set to pre-command.

Add some tests.

Add details to the README.md file for the new hook and provide an example.

jim-barber-he committed:
Add hook parameter

Add the ability to run this as either an environment hook or pre-command
hook with the former being the default.
The behaviour is chosen via a new hook parameter.

I tried switching from the cultureamp/aws-assume-role plugin to this
one but found steps using the artifacts plugin were failing. It is
because this plugin was running before the artifacts had a chance to
run, and so the role being assumed was not able to access our artifacts
S3 bucket. The plugin we switched from ran as a pre-command hook and
didn't suffer from this problem.

Move the code for the plugin from hooks/environment to a shared
lib/plugin.bash script.
Update hooks/environment to call the shared code if the new hook
parameter is unset (making it the default) or set to environment.
Create hooks/pre-command to call the shared code if the new hook
parameter is set to pre-command.

Add some tests.

Add details to the README.md file for the new hook and provide an
example.

jim-barber-he commented:
Force push was sorting our merge conflicts due to changes in the main branch.

jim-barber-he commented:
The latest force push was because some of the changes I made to the README.md file were done by other merged changes performed by others since with minor differences (fix of parameter name, and version bump), so I removed those bits from mine and updated the description accordingly.

toote commented on hooks/environment:
```diff
@@ -2,79 +2,17 @@
-  echo "Using region: ${AWS_REGION}"
-fi
+# shellcheck source=lib/plugin.bash
+. "$DIR/../lib/plugin.bash"
```
I don't think that it is a good idea to source a file and not actually use it (meaning it was sourced for its side effects). If it is a script to be run, execute it directly, otherwise it should define functions to call from this script

toote commented on hooks/pre-command:
```diff
@@ -0,0 +1,17 @@
+DIR="$(cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd)"
+
+# shellcheck source=lib/plugin.bash
+. "$DIR/../lib/plugin.bash"
```
same here: I don't think that it is a good idea to source a file and not actually use it (meaning it was sourced for its side effects). If it is a script to be run, execute it directly, otherwise it should define functions to call from this script

toote commented on tests/environment.bats:
```diff
@@ -47,6 +47,28 @@ run_test_command() {
   unstub buildkite-agent
 }
 
+@test "calls aws sts and exports AWS_ env vars in pre-command hook" {
+  export BUILDKITE_JOB_ID="job-uuid-42"
+  export BUILDKITE_PLUGIN_AWS_ASSUME_ROLE_WITH_WEB_IDENTITY_HOOK="pre-command"
+  export BUILDKITE_PLUGIN_AWS_ASSUME_ROLE_WITH_WEB_IDENTITY_ROLE_ARN="role123"
+
+  stub buildkite-agent "oidc request-token --audience sts.amazonaws.com * : echo 'buildkite-oidc-token'"
+  stub aws "sts assume-role-with-web-identity --role-arn role123 --role-session-name buildkite-job-job-uuid-42 --web-identity-token buildkite-oidc-token : cat tests/sts.json"
+
+  run run_test_command $PWD/hooks/pre-command
+
+  assert_success
+  assert_output --partial "Role ARN: role123"
+  assert_output --partial "Assumed role: assumed-role-id-value"
+
+  assert_output --partial "TESTRESULT:AWS_ACCESS_KEY_ID=access-key-id-value"
+  assert_output --partial "TESTRESULT:AWS_SECRET_ACCESS_KEY=secret-access-key-value"
+  assert_output --partial "TESTRESULT:AWS_SESSION_TOKEN=session-token-value"
+
+  unstub aws
+  unstub buildkite-agent
+}
+
```
What is this supposed to be testing?

jim-barber-he replied:
It's the same as the test above it, but invoking the `pre-command` hook instead of the `environment` hook.

toote replied:
🤦  so sorry I missed that

toote reviewed (changes requested):
Took me a while to understand that your issue was that the artifacts plugin needed to run **before** the role was assumed for the step's command. That said, I understand you refactored everything that was in the hook to its own file but I believe that sourcing it to execute it is a bad practice, just execute it instead. Also, can't understand what the additional test is doing

jim-barber-he committed:
Run the plugin script instead of sourcing it

jim-barber-he committed:
Fix README.md

Something went wrong in one of my previous rebases by the looks and I
lost the closing markdown to the code-snippet example.

toote commented on hooks/environment:
```diff
@@ -2,79 +2,16 @@
-  export "${CREDENTIAL_NAME_PREFIX}AWS_SESSION_TOKEN=${SESSION_TOKEN}"
+HOOK_PARAM="${BUILDKITE_PLUGIN_AWS_ASSUME_ROLE_WITH_WEB_IDENTITY_HOOK:-}"
+if [[ -z "$HOOK_PARAM" || "$HOOK_PARAM" == "environment" ]]; then
+	echo "Running in the environment hook"
```
You should make this code the same as `pre-command` for consistency:

```suggestion
if [[ "${BUILDKITE_PLUGIN_AWS_ASSUME_ROLE_WITH_WEB_IDENTITY_HOOK:-environment}" == "environment" ]]; then	echo "Running in the environment hook"
```

jim-barber-he replied:
That's nicer for sure. Thanks.

toote commented on tests/environment.bats:
```diff
@@ -226,4 +248,14 @@ EOF
 
   unstub aws
   unstub buildkite-agent
+
```
took me a while to figure out the really weird error in tests was due to a missing `}` here:

```suggestion
}
```

toote reviewed (changes requested):
A minor comment and a big issue with tests. I also have a few ideas to refactor some tests but can leave that for a separate PR instead of burden you with it.

jim-barber-he committed:
Code review changes

Make the environment hook code more closely match the pre-environment
hook.

Fix a missing closing brace I lost in the tests when previously
rebasing against the main branch.

toote commented on hooks/pre-command:
```diff
@@ -0,0 +1,16 @@
+
+DIR="$(cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd)"
+
+"$DIR/../lib/plugin.bash"
```
based on the latest test failures, everything indicates you were right to source this in the first place and I led you astray... my bad 😞

toote commented on hooks/environment:
```diff
@@ -2,79 +2,15 @@
-  export AWS_REGION="${AWS_DEFAULT_REGION}"
-  echo "Using region: ${AWS_REGION}"
-fi
+"$DIR/../lib/plugin.bash"
```
based on the latest test failures, everything indicates you were right to source this in the first place and I led you astray... my bad 😞

toote reviewed (changes requested):
From my tests, once these two changes are taken care of, this would be good to go

jim-barber-he committed:
Source plugin code again

jim-barber-he commented:
Once it's all done, will you need me to squash it all up and clean the commit messages?
Or will you do that as part of the merge?

toote commented on lib/plugin.bash:
```diff
@@ -0,0 +1,80 @@
+
+DIR="$(cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd)"
+
+# shellcheck source=shared.bash
```
this should still be referenced as relative to the repository's root:

```suggestion
# shellcheck source=lib/shared.bash
```

toote reviewed (commented):
This should fix the shellcheck error.

toote commented:
> Once it's all done, will you need me to squash it all up and clean the commit messages? Or will you do that as part of the merge?

While not completely against the practice, I prefer to keep the history as-is and do a simple merge commit. Once all tests pass and the PR is approved, your work should be done :)

For those interested in my soapbox: I think that doing additional work for posterity to think that we got it right on the first try is lying to ourselves and could be setting unreasonable standards for those that use these repositories and their histories to learn about the project and its development practices

jim-barber-he committed:
Shellcheck source is relative to repository root

toote reviewed (approved):
👏 Awesoome job, thanks for putting up with my constant feedback! 👏

jim-barber-he commented:
Thanks for accepting my change.
